### PR TITLE
add config of elasticsearch's user and password

### PR DESCRIPTION
**Pull request by WangXiangUSTC**
Add config of elasticsearch's user and password,
and add judge of HTTP response。

### CODE REVIEW

**Comment by WangXiangUSTC** [comment]
It seems that my understand for error is different from your's

File: elastic/client.go [hunk 1]
```diff
@@ -134,20 +139,33 @@ type BulkResponseItem struct {
...
 
-func (c *Client) Do(method string, url string, body map[string]interface{}) (*Response, error) {
-	bodyData, err := json.Marshal(body)
+func (c *Client) DoRequest(method string, url string, body *bytes.Buffer) (*http.Response, error) {
+	req, err := http.NewRequest(method, url, body)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+	if len(c.User) > 0 && len(c.Password) > 0 {
+		req.SetBasicAuth(c.User, c.Password)
```
Comment (siddontang) [inline elastic/client.go]: can we meet the case that user is not empty but the password is?

File: river/river.go [hunk 2]
```diff
@@ -66,7 +66,8 @@ func NewRiver(c *Config) (*River, error) {
 		return nil, errors.Trace(err)
 	}
 
-	r.es = elastic.NewClient(r.c.ESAddr)
+	r.es = elastic.NewClient(r.c.ESAddr, r.c.ESUser, r.c.ESPassword)
```
Comment (siddontang) [inline river/river.go]: please pass a config structure for client now.

File: river/river_test.go [hunk 3]
```diff
@@ -102,8 +103,8 @@ func (s *riverTestSuite) SetUpSuite(c *C) {
 	s.r, err = NewRiver(cfg)
 	c.Assert(err, IsNil)
 
-	err = s.r.es.DeleteIndex("river")
-	c.Assert(err, IsNil)
+	//err = s.r.es.DeleteIndex("river")
```
Comment (siddontang) [inline river/river_test.go]: why comment these?

**Comment by siddontang** [comment]
CI failed @WangXiangUSTC

**Comment by WangXiangUSTC** [comment]
@siddontang I need to make sure is that right return error when http response bigger than 400. 
And this will return error when you delete or update a doc not exist in elasticsearch.
 If you agree with my opinion, I will modify the test code later.

**Comment by siddontang** [comment]
@WangXiangUSTC 

can we keep the origin test and add another for your big response?

**Comment by siddontang** [comment]
@WangXiangUSTC

I think it is ok to return the error if the status code is not 2xx.

**Comment by siddontang** [comment]
Any update? @WangXiangUSTC

**Comment by WangXiangUSTC** [comment]
I modified the test case, and the most of them run success. But the last failure  I don't know why

```
FAIL: client_test.go:98: elasticTestSuite.TestParent 
client_test.go:115:
    c.Assert(err, IsNil)
... value *errors.Err = &errors.Err{message:"", cause:(*errors.Err)(0xc8200166e0), previous:(*errors.Err)(0xc8200166e0), file:"/home/travis/gopath/src/github.com/siddontang/go-mysql-elasticsearch/elastic/client.go", line:207} ("400 Bad Request")

```
It seems that IndexTypeBulk function send a bad request, please  check this.

**Comment by siddontang** [comment]
Hi @EagleChen 

Can you help us to figure it out? I don't use parent feature too.

**Comment by EagleChen** [comment]
@siddontang @WangXiangUSTC 

I will check it sometime today.

**Comment by EagleChen** [comment]
@WangXiangUSTC 

please check [this comment](https://github.com/WangXiangUSTC/go-mysql-elasticsearch/blob/72235fd383dd538a3ea403d7580e07480159885d/elastic/client_test.go#L97)
have you set up the "_mapping" correctly?

you can refer to [this issue](https://github.com/siddontang/go-mysql-elasticsearch/issues/81)  and README on how to set up parent-child relationship

**Comment by WangXiangUSTC** [comment]
ok, I will check tonight  @EagleChen

**Comment by EagleChen** [comment]
@WangXiangUSTC 

I've checked your latest code.
I guess you can create the "parent" index first.

**Comment by siddontang** [comment]
@WangXiangUSTC  CI still failed 😢

**Comment by EagleChen** [comment]
@WangXiangUSTC 

Please reset the last three commits ( from 182549f ). Those changes have nothing to do with the previous bugs.
I think your code at '351742e' is OK.

@siddontang 
I've downloaded 5.x es and run some simple tests locally.
I found that creating index using 'POST' is not valid anymore in "5.x". Which might be the root cause of the failure.
```
$ curl -XPOST localhost:9200/dummy -d @mapping.json
No handler found for uri [/dummy] and method [POST]

$ curl -XPUT localhost:9200/dummy -d @mapping.json
{"acknowledged":true,"shards_acknowledged":true}
```


To you both:
Maybe we can fix the version of elasticsearch to 2.x in travis CI or change the http method to "PUT" at [this line](https://github.com/WangXiangUSTC/go-mysql-elasticsearch/blob/351742e5cc9679dd9527e0b1cf8c8f26830f450b/elastic/client.go#L249) (not tested yet)

**Comment by WangXiangUSTC** [comment]
ok, I reset to  '351742e' already. @EagleChen

**Comment by WangXiangUSTC** [comment]
Hi, @siddontang @EagleChen 
I modify the code, I think this is a big problem.
1. create mapping is for index, not type. I think should not add type in the request url in the function CreateMapping in client.go.  So I modify the code.
2.  When you add parent-child documents, you can't delete the child documents like 
`curl -XDELETE "127.0.0.1:9200/dummy/comment/0"`
this will return:
`{
"error": {
"root_cause": [
{
"type": "routing_missing_exception",
"reason": "routing is required for [dummy]/[comment]/[0]",
"index": "dummy"
}
],
"type": "routing_missing_exception",
"reason": "routing is required for [dummy]/[comment]/[0]",
"index": "dummy"
},
"status": 400
}`
And it should be:
`curl -XDELETE "127.0.0.1:9200/dummy/comment/0?parent=1`
This will return success.
So I comment the delete code in the TestParent.

I think we can accept this pr first, and then fix the bug later.

**Comment by WangXiangUSTC** [comment]
https://www.elastic.co/guide/en/elasticsearch/reference/5.4/docs-delete.html
`Parent
The parent parameter can be set, which will basically be the same as setting the routing parameter.

Note that deleting a parent document does not automatically delete its children. One way of deleting all child documents given a parent’s id is to use the Delete By Query API to perform a index with the automatically generated (and indexed) field _parent, which is in the format parent_type#parent_id.

When deleting a child document its parent id must be specified, otherwise the delete request will be rejected and a RoutingMissingException will be thrown instead.`

**Comment by EagleChen** [comment]
@WangXiangUSTC 

I agree with some of your comments. However, maybe it's a better idea to make the ci pass.

After checking the [ci log](https://travis-ci.org/siddontang/go-mysql-elasticsearch/jobs/239099053),  what I think might work (make 'client_test.go' pass):
1. reset to `351742e`
2. change 'POST' to 'PUT' in [this line](https://github.com/siddontang/go-mysql-elasticsearch/blob/master/elastic/client.go#L219)

Some long explanations are as follows:

1. 
> create mapping is for index, not type. I think should not add type in the request url in the function CreateMapping in client.go. So I modify the code.

'mapping' can be created for a whole index, it can also be created for a specific type. See [this link]
 (https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-put-mapping.html)

I run some simple tests on es 2.2 and es 5.4. Here are what I find:
For es 2.2, 'POST' and 'PUT' both work for index creation:
```
$ curl -XPOST 'localhost:9200/dummy' 
{"acknowledged":true}
$ curl -XPUT 'localhost:9200/dummy1'
{"acknowledged":true}
```
For es 5.4, 'PUT' works but not 'POST':
```
$ curl -XPOST 'localhost:9200/dummy'
No handler found for uri [/dummy] and method [POST]
curl -XPUT 'localhost:9200/dummy'
{"acknowledged":true,"shards_acknowledged":true}
```

After reading `client.go` again, I think in my previous comment, I pointed to the wrong line for the bug. [This line](https://github.com/siddontang/go-mysql-elasticsearch/blob/master/elastic/client.go#L219) might be the root cause.

2. If you read [this code](https://github.com/siddontang/go-mysql-elasticsearch/blob/master/elastic/client.go#L310), you will know the 'parent' feature support is in-complete. But it won't stop 'river' from working correctly.

**Comment by siddontang** [comment]
Hi  @EagleChen 

Can you send us a PR to fix the PUT problem at first? Then we can go on.

**Comment by EagleChen** [comment]
@siddontang 

OK, let me try.

**Comment by siddontang** [comment]
Hi @WangXiangUSTC 

Can you merge the master and run again?

**Comment by WangXiangUSTC** [comment]
@siddontang @EagleChen 
Thanks，ci check success! 
And after merge this request, we can try to fix the bug when delete docs have parent.

**Comment by siddontang** [comment]
HI @WangXiangUSTC @EagleChen 
I think we must send another PR to check the status code first, then we will add the password support. 

It makes more sense.

**Comment by WangXiangUSTC** [comment]
@siddontang you means add more test case? Can you say a little more detail

**Comment by siddontang** [comment]
Hi @WangXiangUSTC 

I say use another PR to check the status code first, then you can the password support later.

I don't want to remove current tests if possible. For password, we can use other new tests especially.

**Comment by WangXiangUSTC** [comment]
@siddontang I will try recently

**Comment by WangXiangUSTC** [comment]
@siddontang Finally I decide to judge http response code after send http request,  instead of return error when code is not 200.

File: elastic/client.go [hunk 4]
```diff
@@ -223,10 +223,13 @@ func (c *Client) CreateMapping(index string, docType string, mapping map[string]
...
 
-	_, err := c.Do("HEAD", reqUrl, nil)
+	r, err := c.Do("HEAD", reqUrl, nil)
+	if err != nil {
+		return errors.Trace(err)
+	}
 
 	// if index doesn't exist, will get 404 not found error, create index first
-	if err != nil {
+	if r.Code != http.StatusOK {
```
Comment (siddontang) [inline elastic/client.go]: should we check 404 here directly?

File: river/sync.go [hunk 5]
```diff
@@ -439,7 +439,7 @@ func (r *River) doBulk(reqs []*elastic.BulkRequest) error {
 	if resp, err := r.es.Bulk(reqs); err != nil {
 		log.Errorf("sync docs err %v after binlog %s", err, r.canal.SyncedPosition())
 		return errors.Trace(err)
-	} else if resp.Errors {
+	} else if resp.Code >= 300 || resp.Code < 200 || resp.Errors {
```
Comment (siddontang) [inline river/sync.go]: `resp.Code / 100 == 2` may be better?

File: river/river_test.go (hunk 3 above)
Comment (siddontang) [inline river/river_test.go]: can we open this case again?

File: river/river_test.go [hunk 6]
```diff
@@ -198,7 +199,11 @@ func (s *riverTestSuite) testElasticGet(c *C, id string) *elastic.Response {
 	docType := "river"
 
 	r, err := s.r.es.Get(index, docType, id)
-	c.Assert(err, IsNil)
+	if err != nil {
+		r := new(elastic.Response)
```
Comment (siddontang) [inline river/river_test.go]: why return a response with `Found=false` here?

**Comment by WangXiangUSTC** [comment]
@siddontang OK， I will check carefully tomorrow.

File: elastic/client.go [hunk 7]
```diff
@@ -214,7 +228,7 @@ func (c *Client) CreateMapping(index string, docType string, mapping map[string]
 		return errors.Trace(err)
 	}
 
-	// index doesn't exist, create index first
+	// if index doesn't exist, will get 404 not found error, create index first
 	if r.Code != http.StatusOK {
```
Comment (siddontang) [inline elastic/client.go]: maybe we can check 404 here directly and return an error for other invalid statuses.

**Comment by WangXiangUSTC** [comment]
@siddontang I have modify the code.

File: elastic/client.go [hunk 8]
```diff
@@ -214,13 +228,15 @@ func (c *Client) CreateMapping(index string, docType string, mapping map[string]
...
-	if r.Code != http.StatusOK {
+	// if index doesn't exist, will get 404 not found, create index first
+	if r.Code == http.StatusNotFound {
 		_, err = c.Do("PUT", reqUrl, nil)
 
 		if err != nil {
 			return errors.Trace(err)
 		}
+	} else if r.Code != http.StatusOK {
+		return errors.Trace(err)
```
Comment (siddontang) [inline elastic/client.go]: seem the error may be nil here.

File: elastic/client.go [hunk 9]
```diff
@@ -214,13 +228,15 @@ func (c *Client) CreateMapping(index string, docType string, mapping map[string]
...
-	if r.Code != http.StatusOK {
+	// if index doesn't exist, will get 404 not found, create index first
+	if r.Code == http.StatusNotFound {
 		_, err = c.Do("PUT", reqUrl, nil)
 
 		if err != nil {
 			return errors.Trace(err)
 		}
+	} else if r.Code != http.StatusOK {
+		return errors.New(resp.Status)
```
Comment (siddontang) [inline elastic/client.go]: I guess compiler may fail here.

**Comment by siddontang** [comment]
Thanks @WangXiangUSTC

**Pull request merged**